test/pin.js

```diff
@@ -0,0 +1,226 @@
+          .then(chunks => {
+            jupiterDir = jupiterDir ||
+              chunks.find(chunk => chunk.path === 'test/fixtures/planets')
+
+            // by separately pinning all the DAG nodes created when adding,
+            // dirs are pinned as type=recursive and
+            // nested pins reference each other
+            return daemon.api.pin.add(chunks.map(chunk => chunk.hash))
+          })
+          .then(() => daemon.api.pin.rm(jupiterDir.hash))
+          .then(() => daemon.api.pin.ls())
+      }
+
+      return withDaemons(pipeline)
+        .then(([goPins, jsPins]) => {
+          expect(goPins.length).to.be.gt(0)
+          expect(goPins).to.deep.include.members(jsPins)
+          expect(jsPins).to.deep.include.members(goPins)
+
+          const dirPin = goPins.find(pin => pin.hash === jupiterDir.hash)
+          expect(dirPin.type).to.eql('indirect')
+        })
+    })
+  })
+
+  describe('ls', function () {
+    it('print same pins', function () {
+      this.timeout(30 * 1000)
+
+      function pipeline (daemon) {
+        return daemon.api.add(jupiter)
+          .then(() => daemon.api.pin.ls())
+      }
+
+      return withDaemons(pipeline)
+        .then(([goPins, jsPins]) => {
+          expect(goPins.length).to.be.gt(0)
+          expect(goPins).to.deep.include.members(jsPins)
+          expect(jsPins).to.deep.include.members(goPins)
+        })
+    })
+  })
+
+  describe(`go and js pinset storage are compatible`, function () {
+    function pipeline (options) {
+      // by starting each daemon with the same repoPath, they
+      // will read/write pins from the same datastore.
+      const repoPath = utils.tmpPath()
+      const content = Buffer.from(String(Math.random()))
+      const pins = []
+
+      return spawnAndStart(options.first, repoPath)
+        .then(daemon => {
+          return daemon.api.add(content)
+            .then(() => daemon.api.pin.ls())
+        })
+        .then(ls => pins.push(ls))
+        .then(() => utils.stopDaemons(daemons))
+        .then(() => spawnAndStart(options.second, repoPath))
+        .then(daemon => daemon.api.pin.ls())
+        .then(ls => pins.push(ls))
+        .then(() => pins)
+    }
+
+    // js-ipfs can read pins stored by go-ipfs
+    // tests that go's pin.flush and js' pin.load are compatible
+    it('go -> js', function () {
+      this.timeout(20 * 1000)
+      this.slow(15000)
+
+      return pipeline({ first: 'go', second: 'js' })
+        .then(([goPins, jsPins]) => {
+          expect(goPins.length).to.be.gt(0)
+          expect(jsPins).to.deep.include.members(goPins)
+          expect(goPins).to.deep.include.members(jsPins)
+        })
+    })
+
+    // go-ipfs can read pins stored by js-ipfs
+    // tests that js' pin.flush and go's pin.load are compatible
+    it.skip('js -> go', function () {
+      // skipped because go can not be spawned on a js repo due to changes in
+      // the DataStore config [link]
```

Comment:
I'd like to understand this better...can `[link]` be a real boy?

Reply:
LOL?